### DISCUSSION TITLE
failed to resize a transparent png keeping bitdepth fixed

### DISCUSSION THREAD

**Discussion by amirian** [ImageMagick/ImageMagick#2031] [Help]
Hi, I'm using IM-7.0.10-13-Q16-x64 to resize a 196KB grayscale png image with transparent background (also made by IM long ago) from 1024x1656 into 1260x2038 pixels. I tried following parameters but gor no correct output: the file gets too large and for PNG00 format the transparency is broken to black. How can I fix this without increasing the bit depth and file size?
convert input.png -resize 1260x2038  1.png                   Output depth:32, size:1.5MB
convert input.png -resize 1260x2038  -depth 4 2.png          Output depth:32, size:830KB
convert input.png -resize 1260x2038  -depth 4 PNG00:3.png    Output depth:8,  size:655KB (Transparency broken, with this warning: "convert: Cannot write image with defined png:bit-depth or png:color-type. `3.png' @ warning/png.c/MagickPNGWarningHandler/1748.")
convert input.png -resize 1260x2038  -depth 2 4.png          Output depth:32, size:430KB
convert input.png -resize 1260x2038  -depth 2 PNG00:5.png    Output depth:4,  size:310KB (Transparency broken)

**Comment by snibgo** [comment]
For v7, I suggest you use "magick", not "convert" (and not "magick convert").

Please show input.png. If it is confidential, please show the complete text output from:
`magick input.png -verbose info:`

**Reply by amirian** [reply]
changing to "magick" didn't affect the output. Is it necessary to show the content of the image? 
magick input.png -verbose info
input.png=>info PNG 1024x1656 1024x1656+0+0 8-bit sRGB 201559B 1.000u 0:00.435
In windows file properties, it is reported as 4 bit depth

**Comment by snibgo** [comment]
Please try again, this time with a colon at the end of info:

**Reply by amirian** [reply]
Image:
  Filename: input.png
  Format: PNG (Portable Network Graphics)
  Mime type: image/png
  Class: DirectClass
  Geometry: 1024x1656+0+0
  Units: Undefined
  Colorspace: sRGB
  Type: GrayscaleAlpha
  Base type: Undefined
  Endianness: Undefined
  Depth: 8-bit
  Channel depth:
    Red: 8-bit
    Green: 8-bit
    Blue: 8-bit
    Alpha: 1-bit
  Channel statistics:
    Pixels: 1695744
    Red:
      min: 0  (0)
      max: 238 (0.933333)
      mean: 45.2599 (0.17749)
      standard deviation: 89.8903 (0.352511)
      kurtosis: 0.297608
      skewness: 1.50363
      entropy: 0.391504
    Green:
      min: 0  (0)
      max: 238 (0.933333)
      mean: 45.2599 (0.17749)
      standard deviation: 89.8903 (0.352511)
      kurtosis: 0.297608
      skewness: 1.50363
      entropy: 0.391504
    Blue:
      min: 0  (0)
      max: 238 (0.933333)
      mean: 45.2599 (0.17749)
      standard deviation: 89.8903 (0.352511)
      kurtosis: 0.297608
      skewness: 1.50363
      entropy: 0.391504
    Alpha:
      min: 0  (0)
      max: 255 (1)
      mean: 80.9071 (0.317283)
      standard deviation: 118.682 (0.465419)
      kurtosis: -1.3835
      skewness: 0.785174
      entropy: 0.901402
  Image statistics:
    Overall:
      min: 0  (0)
      max: 255 (1)
      mean: 54.1717 (0.212438)
      standard deviation: 97.0882 (0.380738)
      kurtosis: -0.244904
      skewness: 1.30708
      entropy: 0.518978
  Alpha: none   #00000000
  Colors: 8
  Histogram:
    1.15771e+006: (0,0,0,0) #00000000 none
    189938: (0,0,0,255) #000000FF black
    5760: (50,50,50,255) #323232FF srgba(50,50,50,1)
    189: (68,68,68,255) #444444FF srgba(68,68,68,1)
    11082: (187,187,187,255) #BBBBBBFF srgba(187,187,187,1)
    82180: (204,204,204,255) #CCCCCCFF grey80
    95438: (221,221,221,255) #DDDDDDFF srgba(221,221,221,1)
    153443: (238,238,238,255) #EEEEEEFF srgba(238,238,238,1)
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 1024x1656+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: Zip
  Orientation: Undefined
  Convex hull: 264,23 760,23 1004,74 1005,75 1013,307 1013,1344 1007,1550 1004,1559 983,1582 982,1583 883,1625 762,1632 263,1632 11,1611 10,1610 10,290 15,70 17,36 155,26 264,23
  Minimum bounding box: 1013,23 1013,1632 10,1632 10,23
  Properties:
    date:create: 2020-05-18T13:57:03+00:00
    date:modify: 2011-10-12T18:40:36+00:00
    minimum-bounding-box:_p: 264,23
    minimum-bounding-box:_q: 760,23
    minimum-bounding-box:_v: 762,1632
    minimum-bounding-box:angle: 0
    minimum-bounding-box:area: 1.61383e+006
    minimum-bounding-box:height: 1003
    minimum-bounding-box:unrotate: -0
    minimum-bounding-box:width: 1609
    png:IHDR.bit-depth-orig: 4
    png:IHDR.bit_depth: 4
    png:IHDR.color-type-orig: 3
    png:IHDR.color_type: 3 (Indexed)
    png:IHDR.interlace_method: 0 (Not interlaced)
    png:IHDR.width,height: 1024, 1656
    png:PLTE.number_colors: 8
    png:sRGB: intent=0 (Perceptual Intent)
    png:tRNS: chunk was found
    signature: 649ee651c26c1cb42d0b1658348611df37cd0f00b8a2e16a3d68511d7b8c6f96
  Tainted: False
  Filesize: 201559B
  Number pixels: 1.69574M
  Pixels per second: 15.623MP
  User time: 0.219u
  Elapsed time: 0:01.108
  Version: ImageMagick 7.0.10-13 Q16 x64 2020-05-17 http://www.imagemagick.org

**Reply by amirian** [reply]
No solution yet?

**Reply by fmw42** [reply]
Can you post an example image that fails?

**Reply by amirian** [reply]
![input](https://user-images.githubusercontent.com/1963455/82438209-ef1e9500-9aad-11ea-945e-75bc437195e5.png)
this is the info:
magick input.png -verbose info:
Image:
  Filename: input.png
  Format: PNG (Portable Network Graphics)
  Mime type: image/png
  Class: DirectClass
  Geometry: 1024x1656+0+0
  Units: Undefined
  Colorspace: sRGB
  Type: GrayscaleAlpha
  Base type: Undefined
  Endianness: Undefined
  Depth: 8/4-bit
  Channel depth:
    Red: 4-bit
    Green: 4-bit
    Blue: 4-bit
    Alpha: 1-bit
  Channel statistics:
    Pixels: 1695744
    Red:
      min: 0  (0)
      max: 238 (0.933333)
      mean: 21.4482 (0.0841105)
      standard deviation: 65.7974 (0.258029)
      kurtosis: 5.90919
      skewness: 2.79778
      entropy: 0.235651
    Green:
      min: 0  (0)
      max: 238 (0.933333)
      mean: 21.4482 (0.0841105)
      standard deviation: 65.7974 (0.258029)
      kurtosis: 5.90919
      skewness: 2.79778
      entropy: 0.235651
    Blue:
      min: 0  (0)
      max: 238 (0.933333)
      mean: 21.4482 (0.0841105)
      standard deviation: 65.7974 (0.258029)
      kurtosis: 5.90919
      skewness: 2.79778
      entropy: 0.235651
    Alpha:
      min: 0  (0)
      max: 255 (1)
      mean: 40.0797 (0.157175)
      standard deviation: 92.8113 (0.363966)
      kurtosis: 1.54881
      skewness: 1.88383
      entropy: 0.627509
  Image statistics:
    Overall:
      min: 0  (0)
      max: 255 (1)
      mean: 26.1061 (0.102377)
      standard deviation: 72.5509 (0.284513)
      kurtosis: 4.4478
      skewness: 2.52374
      entropy: 0.333615
  Alpha: none   #00000000
  Colors: 8
  Histogram:
    1.42922e+006: (0,0,0,0) #00000000 none
    95491: (0,0,0,255) #000000FF black
    11433: (51,51,51,255) #333333FF grey20
    1381: (68,68,68,255) #444444FF srgba(68,68,68,1)
    3141: (187,187,187,255) #BBBBBBFF srgba(187,187,187,1)
    32503: (204,204,204,255) #CCCCCCFF grey80
    41082: (221,221,221,255) #DDDDDDFF srgba(221,221,221,1)
    81498: (238,238,238,255) #EEEEEEFF srgba(238,238,238,1)
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 1024x1656+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: Zip
  Orientation: Undefined
  Convex hull: 507,7 521,8 522,9 814,360 954,530 1023,718 1023,939 975,1138 971,1144 522,1653 507,1655 505,1655 3,1269 2,1268 1,922 1,921 4,729 5,714 80,524 209,352 506,8 507,7
  Minimum bounding box: 512.406,0.580213 1341,716 496,1695 -332,980
  Properties:
    date:create: 2020-05-20T10:51:15+00:00
    date:modify: 2011-10-12T19:21:40+00:00
    minimum-bounding-box:_p: 209,352
    minimum-bounding-box:_q: 506,8
    minimum-bounding-box:_v: 971,1144
    minimum-bounding-box:angle: 40.8078
    minimum-bounding-box:area: 1.41499e+006
    minimum-bounding-box:height: 1292.99
    minimum-bounding-box:unrotate: -40.8078
    minimum-bounding-box:width: 1094.35
    png:IHDR.bit-depth-orig: 4
    png:IHDR.bit_depth: 4
    png:IHDR.color-type-orig: 3
    png:IHDR.color_type: 3 (Indexed)
    png:IHDR.interlace_method: 0 (Not interlaced)
    png:IHDR.width,height: 1024, 1656
    png:PLTE.number_colors: 8
    png:sRGB: intent=0 (Perceptual Intent)
    png:tRNS: chunk was found
    signature: ba1b3588994288366b6f9344b80fb66e69fc2176ff766c32eea75092120fb4b3
  Tainted: False
  Filesize: 113310B
  Number pixels: 1.69574M
  Pixels per second: 18.2991MP
  User time: 0.172u
  Elapsed time: 0:01.092
  Version: ImageMagick 7.0.10-13 Q16 x64 2020-05-17 http://www.imagemagick.org

**Comment by snibgo** [comment]
@amirian: your input has eight colours. Resizing will greatly increase the number of colours, including anti-aliasing of alpha, so the result won't compress so easily, so the output file will be large. You could use "-sample" instead of "-resize", but the result will be fairly horrible.

I don't know what is causing your broken transparency problem. Why do you want "png00:"? This seems to be a bad idea.

You want to resize an image that has already been heavily processed. It would be better to get a high-quality version of that image, and process that to your needs.

**Reply by amirian** [reply]
Thanks for reply. I first tried to re-generate 1260x2038  from scratch, but the old scripts needs a large change to fit with the current IM version. the -sample really lowers the quality. I use png00 only for keeping the bit depth! So, how can I fit the output large png into 4-bit depth without that?

**Comment by snibgo** [comment]
How can you resize that heavily processed image, AND keep reasonable qualiy, AND store it in 4-bits per pixel?  I haven't tried all possible combinations (eg with "-colors 16" and "-depth 4"), but I don't think that is possible.

I suggest you start with a better quality input.

**Reply by amirian** [reply]
I don't think that is the problem, because even with -sample I get a 32 bit output, also even with -colors 16 and -depth 4!

**Comment by fmw42** [comment]
This works OK for me on IM 7.0.10.13 Q16 Mac OSX.

`magick pattern.png -resize 1260x2038 -depth 4 x.png`

![x](https://user-images.githubusercontent.com/13458218/82474731-5a517280-9a80-11ea-933a-6d67df99481c.png)

**Reply by amirian** [reply]
OK, I compared your output "x.png" with the input "pattern.png" on Windows 10. The output number of pixels is 1.5 times of input, but the file size has grown 4 times and the depth becomes 32 (although both shows 8 bit depth with -verbose info of IM).
![screenshot](https://user-images.githubusercontent.com/1963455/82476353-23f90f00-9ae3-11ea-9f2f-e5616b040692.png)

**Reply by fmw42** [reply]
You are comparing images of different dimensions. Try again with the same dimensions. x.png is 1260x238 and your other image is only 1024x1656. So x.png will be larger file size because it is a bigger image.

Try

`magick pattern.png -resize 1024x1656 -depth 4 -type grayscalealpha x.png`

![x](https://user-images.githubusercontent.com/13458218/82478182-6429a480-9a85-11ea-91cd-a0de862a0d52.png)

Filesize: 192549B

**Reply by amirian** [reply]
I edited my reply, the output is only 1.5 times bigger, but the file size becomes 4 times bigger. BTW, the last command also generates a file 1.9 times bigger, yet with 32 bits depth.

**Reply by fmw42** [reply]
Try

`magick pattern.png -strip -resize 1024x1656 -depth 4 -type grayscalealpha -define png:bit-depth=4 PNG8:x.png`

![x](https://user-images.githubusercontent.com/13458218/82480719-10b95580-9a89-11ea-9a56-56d6903e514f.png)

Remaining file size differences are likely from the PNG compression. You can play with that if  you  want. see 

https://imagemagick.org/Usage/formats/#png_write
https://imagemagick.org/script/command-line-options.php#quality

**Reply by amirian** [reply]
That works for the same pixel size, but for enlarging to 1260x2038, I get the following error: 
magick: Invalid palette length `x2.png' @ error/png.c/MagickPNGErrorHandler/1715.

**Reply by fmw42** [reply]
I get the same error message. One of the IM developers will need to look into this.

**Reply by snibgo** [reply]
On the error message, see https://www.imagemagick.org/script/command-line-options.php#define png:bit-depth . To fix this, use "-colors 16".

**Reply by amirian** [reply]
That all works, thank you.
